src/System.Private.CoreLib/src/System/Array.cs

```diff
@@ -897,20 +897,26 @@ internal unsafe static Array NewMultiDimArray(EETypePtr eeType, int * pLengths,
             Debug.Assert(eeType.IsArray && !eeType.IsSzArray);
             Debug.Assert(rank == eeType.ArrayRank);
 
+            ulong totalLength = 1;
+            bool maxArrayDimensionLengthOverflow = false;
+
             for (int i = 0; i < rank; i++)
             {
-                if (pLengths[i] < 0)
+                int length = pLengths[i];
+                if (length < 0)
                     throw new OverflowException();
+                if (length > MaxArrayLength)
+                    maxArrayDimensionLengthOverflow = true;
+                totalLength = totalLength * (ulong)length;
+                if (totalLength > UInt32.MaxValue)
```

Comment:
This should check against `Int32.MaxValue` because of we are casting `totalLength` to `int` below.

The CLR/CoreCLR uses unsigned int for number of elements in multi-dim array, ie it can do `2^32` elements, but CoreRT uses signed int today and thus it can only do `2^31` elements.